### PR TITLE
CLOUDP-211581: Improve memory consumption in Atlas SDK

### PR DESCRIPTION
**Pull request by fmenezes**
## Description

Improve memory consumption in Atlas SDK (by not reading response.Body when we don't need)

Link to any related issue(s): CLOUDP-211581

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Required Checklist:

- [ ] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added any necessary documentation (if appropriate)
- [ ] I have run `make fmt` and formatted my code

### CODE REVIEW

**Comment by fmenezes** [comment]
running tests at https://github.com/mongodb/mongodb-atlas-cli/pull/2979

File: githooks/pre-commit
Comment (wtrocki) [inline githooks/pre-commit:7]: ❤️

File: .github/workflows/autoupdate-preview.yaml
Comment (wtrocki) [reply .github/workflows/autoupdate-preview.yaml:16]: Basically [super nit] I think we still need it locally (as part of the git hooks - tools) and for generation. 
Once we add it there would not be need for any autoupdate workflow changes. 

Those are quite hard to test locally and might fail after merge.

**Comment by wtrocki** [comment]
Verified locally:

1. Creation of AWS cluster
```
▶ go run ./aws_cluster/aws.go
SDK: Created new MongoDB cluster: 665d9c2735f85255276102f9 
Please wait up to 10 minutes for cluster to provision.
```
2. Getting file from backend

File: admin/api_atlas_search.go
Comment (wtrocki) [inline admin/api_atlas_search.go:345]: ❤️

File: admin/api_atlas_search.go
Comment (wtrocki) [inline admin/api_atlas_search.go:351]: [Question] Why closing only on error. Should each method be responsible for closing body independently from if decode succeeded or not.

**Comment by wtrocki** [comment]
When testing error case I noticed that resp.Body is not returned as pointer. 
I modified aws.go to cause error by providing wrong region:

```
go run ./aws_cluster/aws.go  
{0x1400024a180}
https://cloud-dev.mongodb.com/api/atlas/v2/groups/65bd48ebe4cb1d1533f14fd6/clusters POST: HTTP 400 Bad Request (Error code: "INVALID_ATTRIBUTE") Detail: Invalid attribute regionName specified. Reason: Bad Request. Params: [regionName]
2024/06/03 12:57:28 Error when performing SDK request: Invalid attribute regionName specified.
exit status 1
````

On master that renders:

```
{{"detail":"Invalid attribute regionName specified.","error":400,"errorCode":"INVALID_ATTRIBUTE","parameters":["regionName"],"reason":"Bad Request"}}
https://cloud-dev.mongodb.com/api/atlas/v2/groups/65bd48ebe4cb1d1533f14fd6/clusters POST: HTTP 400 Bad Request (Error code: "INVALID_ATTRIBUTE") Detail: Invalid attribute regionName specified. Reason: Bad Request. Params: [regionName]
2024/06/03 13:01:38 Error when performing SDK request: Invalid attribute regionName specified.
exit status 1
```

Is there way to keep previous behaviour?

**Comment by wtrocki** [comment]
@fmenezes Awesome work on making this major improvement!

I have done some testing and left couple questions, but everything looks LGTM. 
We need to clarify possibility of breaking change (as this might require some additional steps for that PR.

File: .github/workflows/autoupdate-preview.yaml
Comment (fmenezes) [reply .github/workflows/autoupdate-preview.yaml:16]: it is already in `make tools`

File: githooks/pre-commit
Comment (fmenezes) [reply githooks/pre-commit:7]: goimports is needed because some imports have to be removed after generation to avoid compilation errors (e.g. `io`)

File: admin/api_atlas_search.go
Comment (fmenezes) [reply admin/api_atlas_search.go:351]: close only happens in this case because it already happens in decode and makeApiError

**Comment by fmenezes** [comment]
> When testing error case I noticed that resp.Body is not returned as pointer. I modified aws.go to cause error by providing wrong region:
> 
> ```
> go run ./aws_cluster/aws.go  
> {0x1400024a180}
> https://cloud-dev.mongodb.com/api/atlas/v2/groups/65bd48ebe4cb1d1533f14fd6/clusters POST: HTTP 400 Bad Request (Error code: "INVALID_ATTRIBUTE") Detail: Invalid attribute regionName specified. Reason: Bad Request. Params: [regionName]
> 2024/06/03 12:57:28 Error when performing SDK request: Invalid attribute regionName specified.
> exit status 1
> ```
> 
> On master that renders:
> 
> ```
> {{"detail":"Invalid attribute regionName specified.","error":400,"errorCode":"INVALID_ATTRIBUTE","parameters":["regionName"],"reason":"Bad Request"}}
> https://cloud-dev.mongodb.com/api/atlas/v2/groups/65bd48ebe4cb1d1533f14fd6/clusters POST: HTTP 400 Bad Request (Error code: "INVALID_ATTRIBUTE") Detail: Invalid attribute regionName specified. Reason: Bad Request. Params: [regionName]
> 2024/06/03 13:01:38 Error when performing SDK request: Invalid attribute regionName specified.
> exit status 1
> ```
> 
> Is there way to keep previous behaviour?

TL;DR we have to change our examples

Basically in our `handleErr` function we `fmt.Println(resp.Body)` instead we should have been printing `err.body` from generic error.

this is misleading because we can't tell what is the state of `resp.Body` after the body was already read.

**Comment by wtrocki** [comment]
> Basically in our handleErr function we fmt.Println(resp.Body) instead we should have been printing err.body from generic error. this is misleading because we can't tell what is the state of resp.Body after the body was already read.

So basically we need documentation to not rely on the resp.Body for printing body from request and follow your advice. 
If customers use that they might need to adjust their changes. 
If that is correct we might want

- Short note in error handling documentation (./docs folder)
- Breaking change note  - no need to make it here (we can reuse existing breaking change update PR as follow up
- Update example

No strong opinion on those.

**Comment by wtrocki** [comment]
Will LGTM once we address the comments. I'm ok to take follow up documentation actions if you want.

**Review by wtrocki — APPROVED** [review]

**Pull request merged**
